### PR TITLE
[terra-time-input] - Keyboard Navigation and Visual Input focus issues

### PR DESCRIPTION
**Pull request by PK106552**
### Summary

**What was changed:**

1. `+` / `-` keys should change the value of the respective field on which the focus is there rather than always updating the minute/second field.
2. Added black dashed outline for time input field.
3. Fixed time input label value to 12 hours format when we have `FORMAT_12_HOUR`.
4. Updated screen reader response to announce hot keys suggestions.

**Why it was changed:**

1. Minute field is incrementing/decrementing instead of second field when we press `+` / `-` keys.
2. Time input filed has blue focus border.
3. Time input label value was 24 hours format when we have `FORMAT_12_HOUR`.

### Testing

This change was tested using:

- [x] WDIO
- [ ] Jest
- [x] Visual testing (please attach a screenshot or recording)
- [ ] Other (please describe below)
- [ ] No tests are needed

### Reviews

In addition to engineering reviews, this PR needs:

- [ ] UX review
- [x] Accessibility review
- [ ] Functional review

### Additional Details

**This PR resolves:**

UXPLATFORM-10116

---

Thank you for contributing to Terra.
@cerner/terra

**Before fix**

https://github.com/cerner/terra-framework/assets/127083404/f8531d91-3501-487f-ae5e-b7facc2f2fe9

Screenshot 2024-01-05 at 1 02 16 PM
Screenshot 2024-01-05 at 1 02 49 PM

**After fix**

Screenshot 2024-01-05 at 1 01 44 PM
Screenshot 2024-01-05 at 1 03 06 PM
Screenshot 2024-01-09 at 3 38 35 PM

**AXE DevTool result**

Screenshot 2024-01-19 at 10 40 57 AM

### CODE REVIEW

**Comment by saket2403** [comment]
I think the +/- keys should increment/decrement the field that is in focus, instead of the seconds or minutes(when the seconds field is not rendered) field every time. Similar to the up and down arrow keys.

**Comment by saket2403** [comment]
The black focus border should also be visible on the AM/PM button. In fact there is no focus indication and it is confusing as a consumer.

https://github.com/cerner/terra-framework/assets/42207428/fedb6f0d-df8e-45f0-8936-4a414af9adf6

**Comment by PK106552** [comment]
> I think the +/- keys should increment/decrement the field that is in focus, instead of the seconds or minutes(when the seconds field is not rendered) field every time. Similar to the up and down arrow keys.

We had discussed about this feature, +/- should increment/decrement the seconds or minutes(when the seconds field is not rendered) field and up/down arrow should increment/decrement the field that is in focus

**Comment by PK106552** [comment]
> The black focus border should also be visible on the AM/PM button. Currently there is no focus indication and it is confusing as a consumer.
> 
>  Screen.Recording.2024-01-10.at.2.28.56.PM.mov

Updated focus indication for AM/PM button.
Screenshot 2024-01-10 at 5 34 02 PM

File: packages/terra-time-input/src/TimeInput.jsx [hunk 1]
```diff
@@ -621,7 +525,7 @@ class TimeInput extends React.Component {
     const previousStateValue = stateValue;
     const variant = TimeUtil.getVariantFromLocale(this.props);
 
-    if (event.keyCode === KeyCode.KEY_UP) {
+    if (event.keyCode === KeyCode.KEY_UP || (event.key === '=' || event.key === '+')) {
```
Comment (supreethmr) [inline packages/terra-time-input/src/TimeInput.jsx:528]: Not able to find any test written for new hot keys added !  if not created already, We need to create one for each keys.

Comment (PK106552) [reply packages/terra-time-input/src/TimeInput.jsx:528]: Hotkeys test has been updated - https://github.com/cerner/terra-framework/pull/1966/commits/266e12a9c438f4eba1aa53b6af09614d0f72994a

Comment (PK106552) [reply packages/terra-time-input/src/TimeInput.jsx:528]: Already existing +/- functionality has =/_ check as well, that's why i have included.
https://github.com/cerner/terra-framework/blob/main/packages/terra-time-input/src/TimeInput.jsx#L485
https://github.com/cerner/terra-framework/blob/main/packages/terra-time-input/src/TimeInput.jsx#L533

File: packages/terra-time-input/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,13 @@
 
 ## Unreleased
 
+* Fixed
+  * Fixed time input to incrememnt the seconds input instead of the minute input wihile pressing `+`, `-` keys.
```
[the remark this answers is not in the document]
Comment (PK106552) [reply packages/terra-time-input/CHANGELOG.md]: Updated - https://github.com/cerner/terra-framework/pull/1966/commits/a055b616c62bc8ec5872e5871b7ba06a2c84fad9

**Review by saket2403 — APPROVED** [review]

**Comment by rbsree** [comment]
+1 for Accessibility Review of Terra Time input component.
 The above-mentioned issues:
 
- The "+ / -" key changes the value of the respective field on which the focus is there rather than always updating the minute/second field.
 - The black dashed outline is added for for time input field when focused.
 - The label for the 12-hour format example announcing the details about the time format if PM or AM. 
 - Screen reader now responds to announce hot key suggestions, then the "N" key can be used to set the current focus.

File: packages/terra-time-input/src/TimeInput.jsx [hunk 3]
```diff
@@ -533,8 +586,48 @@ class TimeInput extends React.Component {
     if ((event.key === '=' || event.key === '+') && !this.props.atMaxDate) {
       const currentTimeValue = this.formatHour(hour, meridiem).concat(':', minute).concat(this.props.showSeconds ? ':'.concat(second) : '');
       if (TimeUtil.validateTime(currentTimeValue, this.props.showSeconds)) {
-        if (minute === '59') {
-          minute = '00';
+        if (inputType === 3 && this.props.showSeconds) {
```
Comment (ShettyAkarsh) [inline packages/terra-time-input/src/TimeInput.jsx:589]: what is this value 3 signifies ?

File: packages/terra-time-input/src/TimeInput.jsx [hunk 4]
```diff
@@ -533,8 +586,48 @@ class TimeInput extends React.Component {
...
+          if (second === '59') {
+            second = '00';
+            if (minute === '59') {
```
Comment (ShettyAkarsh) [inline packages/terra-time-input/src/TimeInput.jsx:592]: can we convert these hard coded values to constant ?

File: packages/terra-time-input/src/TimeInput.jsx [hunk 5]
```diff
@@ -533,8 +586,48 @@ class TimeInput extends React.Component {
...
+              minute = TimeUtil.incrementMinute(minute);
+            }
+          } else {
+            second = TimeUtil.incrementSecond(second);
+          }
+        } else if (inputType === 1) {
+          if (minute === '59') {
+            minute = '00';
+            if (hour === '23') {
+              hour = '00';
```
Comment (supreethmr) [inline packages/terra-time-input/src/TimeInput.jsx:616]: what does zero represents in hour field

File: packages/terra-time-input/src/TimeInput.jsx [hunk 6]
```diff
@@ -485,8 +485,48 @@ class TimeInput extends React.Component {
...
+                } else {
+                  meridiem = this.anteMeridiem;
+                }
+              }
+              hour = TimeUtil.decrementHour(hour, variant);
+            }
+          } else {
+            minute = TimeUtil.decrementMinute(minute);
+          }
+        } else if (inputType === 0) {
```
Comment (supreethmr) [inline packages/terra-time-input/src/TimeInput.jsx:529]: Can you simplify the if else block

File: packages/terra-time-input/src/TimeInput.jsx (hunk 3 above)
Comment (PK106552) [reply packages/terra-time-input/src/TimeInput.jsx:589]: 3 signifies second (https://github.com/cerner/terra-framework/blob/main/packages/terra-time-input/src/TimeUtil.js#L421)

File: packages/terra-time-input/src/TimeInput.jsx (hunk 5 above)
Comment (saket2403) [reply packages/terra-time-input/src/TimeInput.jsx:616]: In 24 hr format, 12 am is rep as 00 right?

Comment (PK106552) [reply packages/terra-time-input/src/TimeInput.jsx:616]: Yes, in 24 hr format, 12 am is represents 00
Screenshot 2024-01-23 at 2 33 53 PM

File: packages/terra-time-input/src/TimeInput.jsx (hunk 6 above)
Comment (PK106552) [reply packages/terra-time-input/src/TimeInput.jsx:509]: Updated - https://github.com/cerner/terra-framework/pull/1966/commits/4caa69c649e337786153c7767f908c3a07d605e7

File: packages/terra-time-input/src/TimeInput.jsx (hunk 4 above)
Comment (PK106552) [reply packages/terra-time-input/src/TimeInput.jsx]: Updated - https://github.com/cerner/terra-framework/pull/1966/commits/4caa69c649e337786153c7767f908c3a07d605e7

File: packages/terra-time-input/src/TimeInput.jsx [hunk 7]
```diff
@@ -485,45 +487,43 @@ class TimeInput extends React.Component {
...
-            hour = '23';
-          } else {
-            if (variant === TimeUtil.FORMAT_12_HOUR && hour === '12') {
-              if (meridiem === this.anteMeridiem) {
-                meridiem = this.postMeridiem;
-              } else {
-                meridiem = this.anteMeridiem;
-              }
+        if (inputType === 3 && this.props.showSeconds) {
+          if ((second === TimeUtil.SingleZeroDigit || second === TimeUtil.DoubleZeroDigit)) {
```
Comment (supreethmr) [inline packages/terra-time-input/src/TimeInput.jsx:491]: Nit : ```suggestion
          if (TimeUtil.initialValue.includes(second)) {
```
singleZero and doubleZeroDigit can be added to single array !!

File: packages/terra-time-input/src/orion-fusion-theme/TimeInput.module.scss [hunk 8]
```diff
@@ -39,6 +39,8 @@
     --terra-time-input-focus-ltr-box-shadow: 0 0 8px #a6d9f4;
     --terra-time-input-focus-rtl-box-shadow: 0 0 8px #a6d9f4;
     --terra-time-input-focus-color: #1c1f21;
+    --terra-time-input-outline: 2px dashed #000;
+    --terra-time-input-meridiem-button-focus-border-outline: 2px dashed #000;
```
Comment (supreethmr) [inline packages/terra-time-input/src/orion-fusion-theme/TimeInput.module.scss:43]: We have not used #000 as outline color for low-light theme I think you need compare this value with other components which has got the outline styles.

File: packages/terra-time-input/src/orion-fusion-theme/TimeInput.module.scss [hunk 9]
```diff
@@ -39,6 +39,8 @@
     --terra-time-input-focus-ltr-box-shadow: 0 0 8px #a6d9f4;
     --terra-time-input-focus-rtl-box-shadow: 0 0 8px #a6d9f4;
     --terra-time-input-focus-color: #1c1f21;
+    --terra-time-input-outline: 2px dashed #000;
```
Comment (supreethmr) [inline packages/terra-time-input/src/orion-fusion-theme/TimeInput.module.scss:42]: We are not making dashed outline changes for fusion theme for other components can you refer to other components that has got outline styles

**Comment by supreethmr** [comment]
@PK106552 outline styles seems to be missing when time input is invalid. Was this intentional ?? Other components still gets outline focus indicator for invalid field as well

File: packages/terra-time-input/src/orion-fusion-theme/TimeInput.module.scss (hunk 9 above)
Comment (PK106552) [reply packages/terra-time-input/src/orion-fusion-theme/TimeInput.module.scss:42]: I have verified with other components (terra-tabs), there we are making dashed outline changes for fusion theme as well (https://github.com/cerner/terra-framework/blob/main/packages/terra-tabs/src/orion-fusion-theme/Tab.module.scss#L5)

File: packages/terra-time-input/src/TimeInput.jsx (hunk 7 above)
Comment (PK106552) [reply packages/terra-time-input/src/TimeInput.jsx]: Updated - https://github.com/cerner/terra-framework/pull/1966/commits/4f5861d144599c94974d061f9a0d9d39d5c0f30b

File: packages/terra-time-input/src/orion-fusion-theme/TimeInput.module.scss (hunk 9 above)
Comment (supreethmr) [reply packages/terra-time-input/src/orion-fusion-theme/TimeInput.module.scss:42]: but we have not made similar changes for date-input, terra-button etc and we can also see the same on some of the newly added terra-framework components like terra-compact-interactive-list, terra-flowsheet-data-grid
Unless it is specified by UX team I think we should not update the fusion styles

Comment (PK106552) [reply packages/terra-time-input/src/orion-fusion-theme/TimeInput.module.scss]: Updated - https://github.com/cerner/terra-framework/pull/1966/commits/78f8257e199c07a68c38a6706123a196bdeb5863#diff-3610f3572ad007f8e3762cfe754e680935cd824c5a140ae9ef76e03b82e55259

File: packages/terra-framework-docs/CHANGELOG.md [hunk 10]
```diff
@@ -12,6 +12,7 @@
 
 * Fixed
   * Fixed font size in `terra-theme-provider` examples.
+  * Fixed label value when we have 12 hours time format in `terra-time-input` examples.
```
Comment (sugan2416) [inline packages/terra-framework-docs/CHANGELOG.md:15]: Update with main. This should be the only added entry

Comment (PK106552) [reply packages/terra-framework-docs/CHANGELOG.md]: Updated

**Review by sugan2416 — APPROVED** [review]

**Pull request merged**